shell/common/shell.cc

```diff
@@ -283,42 +300,65 @@ std::unique_ptr<Shell> Shell::Create(
+    const WindowData window_data,
+    Settings settings,
+    fml::RefPtr<const DartSnapshot> isolate_snapshot,
+    const Shell::CreateCallback<PlatformView>& on_create_platform_view,
+    const Shell::CreateCallback<Rasterizer>& on_create_rasterizer,
+    DartVMRef vm) {
   PerformInitializationTasks(settings);
   PersistentCache::SetCacheSkSL(settings.cache_sksl);
 
   TRACE_EVENT0("flutter", "Shell::CreateWithSnapshots");
 
   if (!task_runners.IsValid() || !on_create_platform_view ||
       !on_create_rasterizer) {
-    return nullptr;
+    std::promise<std::unique_ptr<Shell::ShellHolder>> shell_holder_promise;
+    shell_holder_promise.set_value(nullptr);
+    return shell_holder_promise.get_future();
   }
 
   fml::AutoResetWaitableEvent latch;
-  std::unique_ptr<Shell> shell;
+  std::future<std::unique_ptr<Shell::ShellHolder>> shell_holder_future;
+
   fml::TaskRunner::RunNowOrPostTask(
       task_runners.GetPlatformTaskRunner(),
-      fml::MakeCopyable([&latch,                                          //
+      fml::MakeCopyable([&latch,
                          vm = std::move(vm),                              //
-                         &shell,                                          //
+                         &shell_holder_future,                            //
                          task_runners = std::move(task_runners),          //
                          window_data,                                     //
                          settings,                                        //
                          isolate_snapshot = std::move(isolate_snapshot),  //
                          on_create_platform_view,                         //
                          on_create_rasterizer                             //
   ]() mutable {
-        shell = CreateShellOnPlatformThread(std::move(vm),
-                                            std::move(task_runners),      //
-                                            window_data,                  //
-                                            settings,                     //
-                                            std::move(isolate_snapshot),  //
-                                            on_create_platform_view,      //
-                                            on_create_rasterizer          //
-        );
+        shell_holder_future =
+            CreateShellOnPlatformThread(std::move(vm),
+                                        std::move(task_runners),      //
+                                        window_data,                  //
+                                        settings,                     //
+                                        std::move(isolate_snapshot),  //
+                                        on_create_platform_view,      //
+                                        on_create_rasterizer          //
+            );
         latch.Signal();
       }));
   latch.Wait();
-  return shell;
+  return shell_holder_future;
```

Comment:
will not cause dl. because `RunNowOrPostTask` will 
execute immediately instead of post to next frame if current thread is `PlatformThread`.

But there are still some optimization points here。(latch is unnessary，so i move it)